File: app/src/main/java/com/example/ontime/RestoranAdapter.java
package com.example.ontime;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.ontime.data.Restaurant;

import java.util.ArrayList;
import java.util.List;

public class RestoranAdapter extends RecyclerView.Adapter<RestoranAdapter.RestoranViewHolder> implements Filterable {
    private ArrayList<Restaurant.Data> mRestoranList;
    private List<Restaurant.Data> mRestoranListFull;
    private OnItemClickListener mListener;

    public interface OnItemClickListener {
        void onItemClick(int position);
    }


    public void setOnItemClickListener(OnItemClickListener listener) {
        mListener = listener;
    }

    public static class RestoranViewHolder extends RecyclerView.ViewHolder {

        public ImageView mImageView;
        public TextView mTextView1;
        public TextView mTextView2;
        public TextView mTextView3;

        public RestoranViewHolder(@NonNull View itemView, final OnItemClickListener listener) {
            super(itemView);
            mImageView = itemView.findViewById(R.id.restoran_background);
            mTextView1 = itemView.findViewById(R.id.restoran_name);
            mTextView2 = itemView.findViewById(R.id.restoran_price);
            mTextView3 = itemView.findViewById(R.id.restoran_distance_value);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listener != null) {
                        int position = getAdapterPosition();
                        if (position != RecyclerView.NO_POSITION) {
                            listener.onItemClick(position);
                        }
                    }
                }
            });
        }
    }

    public RestoranAdapter(ArrayList<Restaurant.Data> restoranList) {
        mRestoranList = restoranList;
        mRestoranListFull = new ArrayList<>(restoranList);
    }

    @NonNull
    @Override
    public RestoranViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.restoran_item, parent, false);
        RestoranViewHolder rvh = new RestoranViewHolder(v, mListener);
        return rvh;
    }

    @Override
    public void onBindViewHolder(@NonNull RestoranViewHolder holder, int position) {
        Restaurant.Data currentItem = mRestoranList.get(position);

        holder.mImageView.setImageResource(R.drawable.bg5);
        holder.mTextView1.setText(currentItem.getName());
        holder.mTextView2.setText("400");
        holder.mTextView3.setText("4.5");
    }

    @Override
    public int getItemCount() {
        return mRestoranList.size();
    }

    @Override
    public Filter getFilter() {
        return restoranFilter;
    }

    private Filter restoranFilter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            List<Restaurant.Data> filteredList = new ArrayList<>();

            if (constraint == null || constraint.length() == 0) {
                filteredList.addAll(mRestoranListFull);
            } else {
                String filterPattern = constraint.toString().toLowerCase().trim();

                for (Restaurant.Data item : mRestoranListFull) {
                    if (item.getName().toLowerCase().contains(filterPattern)) {
                        filteredList.add(item);
                    }
                }
            }

            FilterResults results = new FilterResults();
            results.values = filteredList;
            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            mRestoranList.clear();
            mRestoranList.addAll((List) results.values);
            notifyDataSetChanged();
        }
    };
}

File: app/src/main/java/com/example/ontime/RestoranAdapter.kt
package com.example.ontime

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.ontime.RestoranAdapter.RestoranViewHolder
import com.example.ontime.data.Restaurant
import java.util.*

class RestoranAdapter(private val mRestoranList: ArrayList<Restaurant.Data>) : RecyclerView.Adapter<RestoranViewHolder>(), Filterable {
    var mRestoranListFull: List<Restaurant.Data> = ArrayList(mRestoranList)

    class RestoranViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var mImageView: ImageView
        var mTextView1: TextView
        var mTextView2: TextView
        var mTextView3: TextView

        init {
            mImageView = itemView.findViewById(R.id.restoran_background)
            mTextView1 = itemView.findViewById(R.id.restoran_name)
            mTextView2 = itemView.findViewById(R.id.restoran_price)
            mTextView3 = itemView.findViewById(R.id.restoran_distance_value)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RestoranViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.restoran_item, parent, false)
        return RestoranViewHolder(v)
    }

    override fun onBindViewHolder(holder: RestoranViewHolder, position: Int) {
        val (_, name) = mRestoranList[position]
        holder.mImageView.setImageResource(R.drawable.bg5)
        holder.mTextView1.text = name
        holder.mTextView2.text = "400"
        holder.mTextView3.text = "4.5"
    }

    override fun getItemCount(): Int {
        return mRestoranList.size
    }

    override fun getFilter(): Filter {
        return restoranFilter
    }

    private val restoranFilter: Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val filteredList: MutableList<Restaurant.Data> = ArrayList()
            if (constraint == null || constraint.length == 0) {
                filteredList.addAll(mRestoranListFull)
            } else {
                val filterPattern = constraint.toString().toLowerCase().trim { it <= ' ' }
                for (item in mRestoranListFull) {
                    if (item.name.toLowerCase().contains(filterPattern)) {
                        filteredList.add(item)
                    }
                }
            }
            val results = FilterResults()
            results.values = filteredList
            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            mRestoranList.clear()
            mRestoranList.addAll(results.values as Collection<Restaurant.Data>)
            notifyDataSetChanged()
        }
    }

}